src/Castle.Core/DynamicProxy/Generators/Emitters/AbstractTypeEmitter.cs

```diff
@@ -303,7 +303,7 @@ public Type[] GetGenericArgumentsFor(MethodInfo genericMethod)
 			var types = new List<Type>();
 			foreach (var genType in genericMethod.GetGenericArguments())
 			{
-				types.Add(name2GenericType[genType.Name].AsType());
+				types.Add(genericTypeParams[genType.GenericParameterPosition].AsType());
```

Comment:
This makes sense. I wonder if it would be a good idea to check if `genericTypeParams` isn't null and the array contains the element before accessing it since it is provided externally by `SetGenericTypeParameters`, then provide a useful exception, so we don't get something useless like the previous `KeyNotFoundException` one.

Have you looked at the previous `GetGenericArgumentsFor` overload to see if that should also be changed?

Reply:
> I wonder if it would be a good idea to check if `genericTypeParams` isn't null and the array contains the element before accessing it since it is provided externally by `SetGenericTypeParameters`

Having an assertion to that effect might be good, but it shouldn't be any more necessary than checking whether `name2GenericType` contains the key `genType.Name`. Note how the only place where `SetGenericTypeParameters` is called from initializes both `genericTypeParams` and populates `name2GenericType`:

https://github.com/castleproject/Core/blob/808ef994eca34399bf46f934fd4b675ae8e4caf1/src/Castle.Core/DynamicProxy/Generators/Emitters/AbstractTypeEmitter.cs#L121

I'll give it some more thought.

Reply:
> Have you looked at the previous `GetGenericArgumentsFor` overload to see if that should also be changed?

Good point. It indeed looks like this might also profit from the same change. I'll look into it.

Reply:
> Have you looked at the previous `GetGenericArgumentsFor` overload to see if that should also be changed?

@jonorossi, it turns out that other overload is most probably unreachable code. See the new commit for details. Without any test cases, it's hard to whether or not it is faulty. But given that it's public and there may in theory be external consumers, I think the safest option is to leave it as is, but mark it `[Obsolete]`.

We have yet another method, `Type GetGenericArgument(string genericArgumentName)`, which precedes the two that we've discussed so far. The code paths surrounding generic type arguments are rather convoluted, and I don't understand them well enough (yet) to know whether or not there needs to be a change here, too.